tensorflow/core/graph/gpu_fusion_pass.cc

```diff
@@ -0,0 +1,710 @@
+/* Copyright 2018 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#ifdef TENSORFLOW_USE_ROCM
+
+#include <algorithm>
+#include <functional>
+#include <list>
+#include <memory>
+#include <queue>
+#include <set>
+#include <unordered_set>
+#include <utility>
+#include <vector>
+#include "tensorflow/core/common_runtime/function.h"
+#include "tensorflow/core/common_runtime/optimization_registry.h"
+#include "tensorflow/core/framework/node_def_util.h"
+#include "tensorflow/core/graph/algorithm.h"
+#include "tensorflow/core/graph/graph.h"
+#include "tensorflow/core/graph/node_builder.h"
+#include "tensorflow/core/lib/core/status.h"
+#include "tensorflow/core/lib/gtl/array_slice.h"
+#include "tensorflow/core/lib/gtl/map_util.h"
+#include "tensorflow/core/lib/hash/hash.h"
+#include "tensorflow/core/lib/strings/str_util.h"
+#include "tensorflow/core/lib/strings/strcat.h"
+#include "tensorflow/core/platform/logging.h"
+#include "tensorflow/core/util/tensor_format.h"
+
+#include "tensorflow/core/graph/gpu_fusion_pass.h"
+
+namespace tensorflow {
+namespace gpu_fusion_pass {
+
+//----------------------------------------------------------------------
+
+// declaration of all the constants that are used in the gpu_fusion_pass
+// namespace
+const int kVlogLevel = 2;
+
+// graph pass grouping for this fusion pass
+const OptimizationPassRegistry::Grouping kROCmFusionPassGrouping =
+    OptimizationPassRegistry::POST_PLACEMENT;
+
+// attribute name strings
+const string kAttr_T = "T";
```

Comment:
@deven-amd in order to help upstream this pass in the future, I think it's better to change these `string` to `static const char* const`. To my understanding upstream TF tends to avoid using `std::string` unless it's used to deal with underlying operating system. Since this pass is purely in TF proper we should either use `char*` or `StringPiece`.

Reply:
will change to `const char*` in next commit